### PR TITLE
support to use pre generated primitives

### LINKED ISSUE #111: add functionality to get primitive from store if not available generate on the fly
## Feature Description
- while doing inference check if the CryptoStore has primitive already generated, if yes then use the primitive to do the operations in the inference else generate on the fly.
- currently we generate primitive on the fly.

### PR DESCRIPTION
**Pull request by aanurraj**
## Description
this PR closes #111  

## Affected Dependencies
None

## How has this been tested?
tested on local and had a significant improvement.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: src/sympc/protocol/spdz/spdz.py [hunk 1]
```diff
@@ -55,24 +55,27 @@ def mul_master(
...
-
     args = [
         list(el) + [op_str]
         for el in zip(session.session_ptrs, x.share_ptrs, y.share_ptrs)
     ]
 
-    mask = parallel_execution(spdz_mask, session.parties)(args)
+    try:
+        mask = parallel_execution(spdz_mask, session.parties)(args)
+    except (KeyError, ValueError):
```
Comment (LaRiffle) [inline src/sympc/protocol/spdz/spdz.py:65]: raise a specific Exception to avoid catching other unexpected errors 
And don't forget to add tests :)

Comment (aanurraj) [reply src/sympc/protocol/spdz/spdz.py]: done

File: src/sympc/exceptions.py [hunk 2]
```diff
@@ -0,0 +1,5 @@
+"""Specific sympc exceptions."""
```
Comment (gmuraru) [inline src/sympc/exceptions.py:1]: Add this in the "sympc.store" location

File: src/sympc/protocol/beaver/beaver.py [hunk 3]
```diff
@@ -250,15 +255,19 @@ def matmul_store_get(
...
     Raises:
-        ValueError: If no primitive in the store for config_key.
+        EmptyPrimitiveStore: If no primitive in the store for config_key.
     """
     config_key = f"beaver_matmul_{tuple(a_shape)}_{tuple(b_shape)}"
-    primitives = store[config_key]
+
+    try:
+        primitives = store[config_key]
+    except Exception:
```
Comment (gmuraru) [inline src/sympc/protocol/beaver/beaver.py:264]: MAYBE: This is the top-level exception, try to be more specific with this

File: src/sympc/protocol/beaver/beaver.py [hunk 4]
```diff
@@ -329,15 +338,19 @@ def conv2d_store_get(
...
 
     Raises:
-        ValueError: If no primitive in the store for config_key.
+        EmptyPrimitiveStore: If no primitive in the store for config_key.
     """
     config_key = f"beaver_conv2d_{tuple(a_shape)}_{tuple(b_shape)}"
-    primitives = store[config_key]
+
+    try:
+        primitives = store[config_key]
```
Comment (gmuraru) [inline src/sympc/protocol/beaver/beaver.py:346]: Also with this

File: src/sympc/protocol/beaver/beaver.py [hunk 5]
```diff
@@ -329,15 +338,19 @@ def conv2d_store_get(
...
+    except Exception:
+        raise EmptyPrimitiveStore(f"{config_key} does not exists in the store")
 
     try:
         primitive = primitives[0]
     except Exception:
```
Comment (gmuraru) [inline src/sympc/protocol/beaver/beaver.py:352]: Also with this

File: tests/sympc/protocol/protocol_test.py [hunk 6]
```diff
@@ -25,3 +29,51 @@ def test_exception_insufficient_fss_primitives():
...
+        conv2d_store_get(store, a_shape, b_shape)
+
+
+def test_exception_empty_primitive_store_no_primitive():
+    """Exception when there exists a key in store but has no primitive."""
+    a_shape = (2, 2)
+    b_shape = (2, 2)
+
+    store = {
+        "beaver_mul_(2, 2)_(2, 2)": [],
```
Comment (gmuraru) [inline tests/sympc/protocol/protocol_test.py:56]: Q: Can't we use ```pytest.parametrize```?

**Comment by gmuraru** [comment]
@aanurraj looks great! Could you also add a new benchmark test where we have the showcases of what happens when we create the primitives before the computation? (ideally, it should be a separate graph with 2 colors -- online vs offline computation of primitives)

**Review by gmuraru — CHANGES_REQUESTED** [review]
Check comments :100:

**Review by gmuraru — APPROVED** [review]
LGTM!

**Pull request merged**